### ISSUE TITLE
Higher level filesystem helpers

### ISSUE THREAD

**Issue by cgwalters** [bytecodealliance/cap-std#211]
Hi, I'm working on porting away from openat to cap-std.  I maintain the [openat-ext crate](https://docs.rs/openat-ext/latest/openat_ext/index.html) that has a lot of "high level" helpers:
https://docs.rs/openat-ext/latest/openat_ext/trait.OpenatDirExt.html

(We can ignore the FileExt trait for now)

Of these, I find myself using e.g. `open_file_optional()` quite often.

But the other bigger and more interesting hunk of functionality is https://docs.rs/openat-ext/latest/openat_ext/trait.OpenatDirExt.html#tymethod.new_file_writer

Basically we have a nice wrapper around `O_TMPFILE` + `linkat`.

Would it make sense to try to add these to [cap-fs-ext](https://docs.rs/cap-fs-ext/latest/cap_fs_ext/) or should I start with a new, separate crate?

**Comment by sunfishcode** [comment]
cap-fs-ext doesn't have a super clear scope yet, so I'm open to ideas here. Absent any better ideas, my sense is that things that need to directly depend on cap-primitives can go in cap-fs-ext, while things that can work in terms of cap-std's public API can go in separate crates. How does that sound?

**Comment by cgwalters** [comment]
OK, makes sense.  Hmm, actually another big axis here is cap-std is trying to be portable to e.g. Windows and WASI, but for large hunks of my code I only care about Linux.

I didn't have this problem before because the `openat` crate only worked on Unix and had some methods like [this one](https://docs.rs/openat/0.1.19/openat/struct.Dir.html#method.new_unnamed_file) that just fail at runtime on non-Linux.  (Which now that I think about it is strange instead of having it be under `#[cfg]`, maybe the crate author hoped *BSD etc. would implement `O_TMPFILE` quickly too)

**Comment by sunfishcode** [comment]
I'm open to having Linux-only APIs in cap-fs-ext, using either `#[cfg]` or failing at runtime, depending on which is appropriate for a given function.

Thinking about this a little more, the "do you need to call cap-primitives yourself" test still feels like the main thing we want for the determining factor as to whether something goes in cap-fs-ext or a different crate.

**Comment by cgwalters** [comment]
Calling this one closed for now by https://github.com/bytecodealliance/cap-std/pull/229

(It's not the end of the story, but an answer for now)
